service/qos/service_level_controller.cc

```diff
@@ -249,7 +268,15 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
     if (sl_it != _service_levels_db.end()) {
         _service_levels_db.erase(sl_it);
```

Comment:
Is this called under some lock? Otherwise different notifications will get mixed up.

Reply:
Yes, at the top level, there is a serialization semaphore (with one units). The global controller grabs it for every notification it sends and releases it when it is done so every shard can handle only one notification simultaneously.
This semaphore is at: `service_level_controller::global_controller_data::notifications_serializer